frame/ranked-collective/src/lib.rs

```diff
@@ -0,0 +1,540 @@
+pub struct MemberRecord {
+	/// The rank of the member.
+	rank: Rank,
+}
+
+/// Record needed for every vote.
+#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum VoteRecord {
+	/// Vote was an aye with given vote weight.
+	Aye(Votes),
+	/// Vote was a nay with given vote weight.
+	Nay(Votes),
+}
+
+impl From<(bool, Votes)> for VoteRecord {
+	fn from((aye, votes): (bool, Votes)) -> Self {
+		match aye {
+			true => VoteRecord::Aye(votes),
+			false => VoteRecord::Nay(votes),
+		}
+	}
+}
+
+/// Vote-weight scheme where all voters get one vote regardless of rank.
+pub struct Unit;
+impl Convert<Rank, Votes> for Unit {
+	fn convert(_: Rank) -> Votes {
+		1
+	}
+}
+
+/// Vote-weight scheme where all voters get one vote plus an additional vote for every excess rank
+/// they have. I.e.:
+///
+/// - Each member with no excess rank gets 1 vote;
+/// - ...with an excess rank of 1 gets 2 votes;
+/// - ...with an excess rank of 2 gets 2 votes;
+/// - ...with an excess rank of 3 gets 3 votes;
+/// - ...with an excess rank of 4 gets 4 votes.
+pub struct Linear;
+impl Convert<Rank, Votes> for Linear {
+	fn convert(r: Rank) -> Votes {
+		(r + 1) as Votes
+	}
+}
+
+/// Vote-weight scheme where all voters get one vote plus additional votes for every excess rank
+/// they have incrementing by one vote for each excess rank. I.e.:
+///
+/// - Each member with no excess rank gets 1 vote;
+/// - ...with an excess rank of 1 gets 2 votes;
+/// - ...with an excess rank of 2 gets 3 votes;
+/// - ...with an excess rank of 3 gets 6 votes;
+/// - ...with an excess rank of 4 gets 10 votes.
+pub struct Geometric;
+impl Convert<Rank, Votes> for Geometric {
+	fn convert(r: Rank) -> Votes {
+		let v = (r + 1) as Votes;
+		v * (v + 1) / 2
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// The outer event type.
+		type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The origin required to add, promote or remove a member.
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The polling system used for our voting.
+		type Polls: Polling<
+			TallyOf<Self, I>,
+			Votes = Votes,
+			Class = Rank,
+			Moment = Self::BlockNumber,
+		>;
+
+		/// Convert a rank_delta into a number of votes the rank gets.
+		///
+		/// Rank_delta is defined as the number of ranks above the minimum required to take part
+		/// in the poll.
+		type VoteWeight: Convert<Rank, Votes>;
+	}
+
+	/// The number of members in the collective who have at least the rank according to the index
```

Comment:
Huh, "at least the rank" and not simply "the same rank as the index"?

Reply:
That's right.

Reply:
Yes. This way you can enumerate and also randomly select from members of *at least* a particular rank.